src/lib/broadcast.jl

```diff
@@ -59,6 +59,11 @@ unbroadcast(x::Union{Number,Ref}, x̄) = accum_sum(x̄)
 @adjoint broadcasted(::typeof(*), x, y) = x.*y,
   z̄ -> (nothing, unbroadcast(x, z̄ .* y), unbroadcast(y, z̄ .* x))
 
+@adjoint broadcasted(::typeof(*), x::Array{<:Number}, y::Array{<:Number}) = x.*y,
+  z̄ -> (nothing, unbroadcast(x, z̄ .* conj(y)), unbroadcast(y, z̄ .* conj(x)))
+@adjoint broadcasted(::typeof(*), x::Number, y::Number) = x.*y,
+  z̄ -> (nothing, unbroadcast(x, z̄ .* conj(y)), unbroadcast(y, z̄ .* conj(x)))
```

Comment:
You may as well fold these definitions into the `Any` definition above and add the `conj` there. Also, best if it's broadcasted, so that it can be fused where possible.

Reply:
I thinkg for user defined `*` operations, using `conj` here may cause trouble? Calling the backward function of `*` with respect to specific types is desired.

Reply:
The only way to do that is to just delete the special case entirely and have it fall back.

It shouldn't matter though; if someone has defined a `*` operation it should behave mathematically like `*`, and if it doesn't have the same gradient I see that as their issue.

Reply:
> if someone has defined a * operation it should behave mathematically like *

This is often not true, maybe an error is better?

I tried to construct a counter example, but the error is not as expected
```julia
julia> using Zygote

julia> struct A
       data::Float64
       end
julia> Base.:*(a::A, b::A) = A(a.data+b.data)

julia> gradient(x->(A(x)*A(x)).data, 0.2)
(2.0,)

julia> gradient(x->([A(x)].*[A(x)])[].data, 0.2)
ERROR: MethodError: no method matching *(::NamedTuple{(:data,),Tuple{Float64}}, ::A)
Closest candidates are:
  *(::Any, ::Any, ::Any, ::Any...) at operators.jl:502
  *(::A, ::A) at REPL[6]:1
Stacktrace:
 [1] _broadcast_getindex_evalf at ./broadcast.jl:578 [inlined]
 [2] _broadcast_getindex at ./broadcast.jl:551 [inlined]
 [3] getindex at ./broadcast.jl:511 [inlined]
 [4] macro expansion at ./broadcast.jl:843 [inlined]
 [5] macro expansion at ./simdloop.jl:73 [inlined]
 [6] copyto! at ./broadcast.jl:842 [inlined]
 [7] copyto! at ./broadcast.jl:797 [inlined]
 [8] copy at ./broadcast.jl:773 [inlined]
 [9] materialize at ./broadcast.jl:753 [inlined]
 [10] (::getfield(Zygote, Symbol("##1054#1055")){Array{A,1},Array{A,1}})(::Array{Any,1}) at /home/leo/.julia/dev/Zygote/src/lib/broadcast.jl:60
 [11] #5 at /home/leo/.julia/dev/Zygote/src/lib/grad.jl:46 [inlined]
 [12] (::typeof(∂(getfield(Main, Symbol("##5#6"))())))(::Float64) at /home/leo/.julia/dev/Zygote/src/compiler/interface2.jl:0
 [13] (::getfield(Zygote, Symbol("##71#72")){typeof(∂(getfield(Main, Symbol("##5#6"))()))})(::Float64) at /home/leo/.julia/dev/Zygote/src/compiler/interface.jl:38
 [14] gradient(::Function, ::Float64) at /home/leo/.julia/dev/Zygote/src/compiler/interface.jl:47
 [15] top-level scope at none:0
```

If I comment the following two lines in `broadcast.jl`
```julia
#@adjoint broadcasted(::typeof(*), x, y) = x.*y,
#  z̄ -> (nothing, unbroadcast(x, z̄ .* y), unbroadcast(y, z̄ .* x))
```
I got
```julia
julia> gradient(x->([A(x)].*[A(x)])[].data, 0.2)
ERROR: ArgumentError: tuple must be non-empty
Stacktrace:
 [1] first(::Tuple{}) at ./tuple.jl:77
 [2] unzip(::Tuple{}) at /home/leo/.julia/dev/Zygote/src/lib/array.jl:76
 [3] adjoint at /home/leo/.julia/dev/Zygote/src/lib/array.jl:83 [inlined]
 [4] _forward at /home/leo/.julia/dev/Zygote/src/lib/grad.jl:44 [inlined]
 [5] broadcasted at ./broadcast.jl:1166 [inlined]
 [6] #17 at ./none:1 [inlined]
 [7] _forward(::Zygote.Context, ::getfield(Main, Symbol("##17#18")), ::Float64) at /home/leo/.julia/dev/Zygote/src/compiler/interface2.jl:0
 [8] _forward(::Function, ::Float64) at /home/leo/.julia/dev/Zygote/src/compiler/interface.jl:31
 [9] forward(::Function, ::Float64) at /home/leo/.julia/dev/Zygote/src/compiler/interface.jl:37
 [10] gradient(::Function, ::Float64) at /home/leo/.julia/dev/Zygote/src/compiler/interface.jl:46
 [11] top-level scope at none:0
```

I believe fixing this is beyond my ability.

Reply:
I think it'd be reasonable to define `Numeric{T<:Number} = Union{T,AbstractArray{T}}` and restrict the broadcast definition to that, rather than the current `Any`, to avoid this kind of problem.

Reply:
As another example could also define `map(A, B) = A*B` but at some point we just have to say ok, you're on your own here. We haven't nailed down exactly what that point is yet, but we can at least rule out certain cases that we're definitely not going to support.